### PR TITLE
get rid of deep copy

### PR DESCRIPTION
**Pull request by tiansijie**

### CODE REVIEW

File: src/infinity-menu.js [hunk 1]
```diff
@@ -53,9 +52,9 @@ export default class InfinityMenu extends React.Component {
 			return true;
 		}
 
-		if (this.state.search.isSearching &&
-			this.state.search.isSearching === nextState.search.isSearching &&
-			this.state.search.searchInput === nextState.search.searchInput) {
+		 if (this.state.search.isSearching
```
Comment (conorhastings) [inline src/infinity-menu.js]: can't we also return false if not searching?

File: src/infinity-menu.js [hunk 2]
```diff
@@ -113,12 +115,16 @@ export default class InfinityMenu extends React.Component {
 				return this.findFilted(p, c, k);
 			}, []) : [];
 			if (filteredSubFolder.length !== 0) {
-				node.isOpen = true;
+				node.isSearchOpen = true;
```
Comment (conorhastings) [inline src/infinity-menu.js]: why this change?

File: src/infinity-menu.js [hunk 3]
```diff
@@ -113,12 +115,16 @@ export default class InfinityMenu extends React.Component {
...
 				node.children = filteredSubFolder;
+				node.isSearchDisplay = true;
 				trees[key] = node;
 				return trees;
 			}
 			else {
+				node.isSearchOpen = false;
+				node.isSearchDisplay = false;
+				trees[key] = node;
```
Comment (conorhastings) [inline src/infinity-menu.js]: what is this doing exactly?

File: src/infinity-menu.js [hunk 4]
```diff
@@ -266,7 +278,7 @@ export default class InfinityMenu extends React.Component {
 	 *  @description React render method for creating infinity menu
 	 */
 	render() {
-		const tree = dcopy(this.props.tree);
+		const tree = this.props.tree;
```
Comment (conorhastings) [inline src/infinity-menu.js]: can you just quickly explain what parts of the changes allowed us to remove deep copy? @tiansijie

File: src/infinity-menu.js (hunk 2 above)
Comment (tiansijie) [inline src/infinity-menu.js]: to make search working correct, we need to have an searchOpen property to mark what is open inside search and not affect the original open status

Comment (conorhastings) [inline src/infinity-menu.js]: but why does isOpen get removed?

File: src/infinity-menu.js (hunk 3 above)
Comment (tiansijie) [inline src/infinity-menu.js]: this is basically reconstruct the tree with new value

**Comment by conorhastings** [comment]
looks like the folders that don't have matching leafs no longer get filtered

File: src/infinity-menu.js (hunk 2 above)
Comment (tiansijie) [inline src/infinity-menu.js]: because this part is only open in search and `isOpen` is open under the non-search

File: src/infinity-menu.js (hunk 3 above)
Comment (conorhastings) [inline src/infinity-menu.js]: 👍

**Comment by tiansijie** [comment]
@conorhastings should fix the searching folder problem

File: src/infinity-menu.js (hunk 4 above)
Comment (tiansijie) [inline src/infinity-menu.js]: So I am adding `isSearchDisplay` and `isSearchOpen` to the tree to get rid of deep copy because searching is the only part we actually deleted things from the tree which was why we originally need the deep copy

File: src/infinity-menu.js (hunk 1 above)
Comment (tiansijie) [inline src/infinity-menu.js]: this statement is only tell us that if search input is same, we don't need to re render

File: src/infinity-menu.js (hunk 4 above)
Comment (conorhastings) [inline src/infinity-menu.js]: what if the user modifies the tree on their end without searching through the component? just not supported ?

File: src/infinity-menu.js [hunk 5]
```diff
@@ -114,7 +114,16 @@ export default class InfinityMenu extends React.Component {
 			const filteredSubFolder = node.children.length ? node.children.reduce((p, c, k) => {
 				return this.findFilted(p, c, k);
```
Comment (conorhastings) [inline src/infinity-menu.js]: should this be findFiltered?

File: src/infinity-menu.js [hunk 6]
```diff
@@ -114,7 +114,16 @@ export default class InfinityMenu extends React.Component {
...
 			}, []) : [];
-			if (filteredSubFolder.length !== 0) {
+			const shouldDisplay = filteredSubFolder.reduce((prev, child) => {
```
Comment (conorhastings) [inline src/infinity-menu.js]: does this loop over all items in a folder? might be better served as a for loop or a `.some` so we don't need to loop over every item once we know we are returning true

File: src/infinity-menu.js (hunk 4 above)
Comment (tiansijie) [inline src/infinity-menu.js]: The `isSearchDisplay` and `isSearchOpen` just we added to the object, even the user modifiles the tree, we should still able to display them correctly

**Comment by conorhastings** [comment]
@tiansijie looks good, sorry this took so long to get in, feel free to version bump.
